File: Our_model/divide.py
import re
import json

def divide_blocks(js_code: str):
    lines = js_code.strip().splitlines()
    blocks = []
    current_block = []
    block_id = 1
    brace_depth = 0

    def flush_block():
        nonlocal current_block, block_id
        if not current_block:
            return

        trimmed = [l.strip() for l in current_block]
        is_only_brace = len(trimmed) == 1 and trimmed[0] == "}"

        if not is_only_brace:
            blocks.append({
                "id": block_id,
                "code": "\n".join(current_block).strip(),
                "successors": []  # Not handled for now
            })
            block_id += 1

        current_block = []

    for i, line in enumerate(lines):
        trimmed = line.strip()
        if not trimmed:
            continue

        # Update brace depth
        brace_depth += len(re.findall(r"{", line))
        brace_depth -= len(re.findall(r"}", line))

        current_block.append(line)

        is_control = re.match(r"^(if|else|for|while|switch|function)\b", trimmed)
        is_ender = re.match(r"^(return|break|continue|throw)\b", trimmed)
        ends_with_semicolon = trimmed.endswith(";")
        next_line = lines[i + 1].strip() if i + 1 < len(lines) else ""

        should_flush = (
            (is_control and brace_depth == 1 and len(current_block) == 1)
            or is_ender
            or ends_with_semicolon
            or (brace_depth == 0 and not next_line)
        )

        if should_flush:
            flush_block()

    # Final flush
    flush_block()

    return blocks


# For debug/test
if __name__ == "__main__":
    import sys

    input_code = sys.stdin.read()
    blocks = divide_blocks(input_code)
    print(json.dumps(blocks, indent=2))

File: Our_model/divide.js
// divide_heuristic.js -- This is to divide only incorrect blocks i.e. buggy code
const readline = require("readline");
let input = '';
process.stdin.setEncoding("utf8");
process.stdin.on("data", chunk => input += chunk);
process.stdin.on("end", () => {
    const lines = input.split("\n");
    const blocks = [];
    let currentBlock = [];
    let blockId = 1;
    let braceDepth = 0;

    const flushBlock = () => {
        if (currentBlock.length === 0) return;

        const trimmedLines = currentBlock.map(l => l.trim());
        const isOnlyBrace = trimmedLines.length === 1 && trimmedLines[0] === "}";

        if (!isOnlyBrace) {
            blocks.push({
                id: blockId++,
                code: currentBlock.join("\n").trim(),
                successors: []
            });
        }

        currentBlock = [];
    };

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        const trimmed = line.trim();
        if (trimmed === '') continue;

        // Update brace depth first
        braceDepth += (line.match(/{/g) || []).length;
        braceDepth -= (line.match(/}/g) || []).length;

        currentBlock.push(line);

        const isControl = /^(if|else|for|while|switch|function)\b/.test(trimmed);
        const isEnder = /^(return|break|continue|throw)\b/.test(trimmed);
        const endsSemicolon = trimmed.endsWith(';');
        const nextLine = lines[i + 1]?.trim();

        const shouldFlush = (
            isControl && braceDepth === 1 && currentBlock.length === 1
        ) || (
            isEnder || endsSemicolon
        ) || (
            braceDepth === 0 && (!nextLine || nextLine === '')
        );

        if (shouldFlush) {
            flushBlock();
        }
    }

    // Flush remaining block
    flushBlock();

    console.log(JSON.stringify(blocks, null, 2));
});
